### PR TITLE
fix: Validate if issuers config was already updated before updating it

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

- We validate that the issuers config is not yet updated before updating it again. This should prevent us from writing to that config more than once avoiding any unexpected behaviour.
- The integration test changes introduced in #386 made the CI unstable. The changes in the issuers config might not be immediately visible in vault, for that in this PR we wait until all units are active idle (that indicates that the Vault API is available again).

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by DanielArndt** [comment]
This might be because `wait_for_idle` is extremely easy to misuse, especially when dealing with multiple applications.

By default, `wait_for_at_least_units` is 1, which means (as I understand) that as soon as _any_ one unit that belongs to the applications listed is "active", it will move forward.

So, we may want to bump the number of units to wait for to (num vault units + 1 for pki requirer) to make sure all units are ready. If we really only need one vault unit to be ready, I think we'd need to split them into two separate waits.

**Comment by saltiyazan** [comment]
> This might be because `wait_for_idle` is extremely easy to misuse, especially when dealing with multiple applications.
> 
> By default, `wait_for_at_least_units` is 1, which means (as I understand) that as soon as _any_ one unit that belongs to the applications listed is "active", it will move forward.
> 
> So, we may want to bump the number of units to wait for to (num vault units + 1 for pki requirer) to make sure all units are ready. If we really only need one vault unit to be ready, I think we'd need to split them into two separate waits.

After some testing, yes that should work, however I don't see right now the connection between the status of the unit and what config is applied in Vault. I propose having both to avoid any unstability in the CI.

**Comment by DanielArndt** [comment]
> I don't see right now the connection between the status of the unit and what config is applied in Vault

If we don't wait for it to settle after calling `vault_app.set_config()`, we may be hitting the API before it has time to consider the changed. Calling `wait_for_idle()` with the PKI requirer app in the parameters means that we don't wait at all because we haven't done anything to require work on the PKI requirer app, so it is already idle and we immediately return.

There may be some cross test issue too, since in the test above we make the same wait call, and we're not actually waiting for both charms to finish dealing with the integration.

> I propose having both to avoid any unstability in the CI.

Personally, I'd rather lean the other direction. Lets not add complication unless we're sure we need it. If we don't definitively know the cause and it seems to work after correcting this `wait_for_idle()` stuff, I'd let it simmer for a bit and see if it re-emerges.

**Comment by saltiyazan** [comment]
> > I don't see right now the connection between the status of the unit and what config is applied in Vault
> 
> If we don't wait for it to settle after calling `vault_app.set_config()`, we may be hitting the API before it has time to consider the changes. Calling `wait_for_idle()` with the PKI requirer app in the parameters means that we don't wait at all because we haven't done anything to require work on the PKI requirer app, so it is already idle and we immediately return.
> 
> There may be some cross test issue too, since in the test above we make the same wait call, and we're not actually waiting for both charms to finish dealing with the integration.
> 
> > I propose having both to avoid any unstability in the CI.
> 
> Personally, I'd rather lean the other direction. Lets not add complication unless we're sure we need it. If we don't definitively know the cause and it seems to work after correcting this `wait_for_idle()` stuff, I'd let it simmer for a bit and see if it re-emerges.

Okay, I see your point, I reverted those changes and kept the ones for waiting for the units to be active idle. We can revisit if the problem occurs.
Can you please take a look? @DanielArndt

**Comment by saltiyazan** [comment]
> > > I don't see right now the connection between the status of the unit and what config is applied in Vault
> > 
> > 
> > If we don't wait for it to settle after calling `vault_app.set_config()`, we may be hitting the API before it has time to consider the changes. Calling `wait_for_idle()` with the PKI requirer app in the parameters means that we don't wait at all because we haven't done anything to require work on the PKI requirer app, so it is already idle and we immediately return.
> > There may be some cross test issue too, since in the test above we make the same wait call, and we're not actually waiting for both charms to finish dealing with the integration.
> > > I propose having both to avoid any unstability in the CI.
> > 
> > 
> > Personally, I'd rather lean the other direction. Lets not add complication unless we're sure we need it. If we don't definitively know the cause and it seems to work after correcting this `wait_for_idle()` stuff, I'd let it simmer for a bit and see if it re-emerges.
> 
> Okay, I see your point, I reverted those changes and kept the ones for waiting for the units to be active idle. We can revisit if the problem occurs. Can you please take a look? @DanielArndt

The CI just failed, I'm not sure if I'm doing something wrong based on your description or if this won't be enough.
Check https://github.com/canonical/vault-k8s-operator/actions/runs/9463126946/job/26068340370?pr=388

**Comment by DanielArndt** [comment]
Hmm... still broken even with the wait included again... so weird. This might be worthy of some additional investigation. It seems like we're not accounting for something.

File: lib/charms/vault_k8s/v0/vault_client.py [hunk 1]
```diff
@@ -469,13 +469,19 @@ def make_latest_pki_issuer_default(self, mount: str) -> None:
...
-        self._client.write_data(
-            path=f"{mount}/config/issuers",
-            data={
-                "default_follows_latest_issuer": True,
-                "default": first_issuer,
-            },
-        )
+        try:
+            issuers_config = self._client.read(path=f"{mount}/config/issuers")
+            if issuers_config and not issuers_config["data"]["default_follows_latest_issuer"]:  # type: ignore -- bad type hint in stubs
```
Comment (DanielArndt) [inline lib/charms/vault_k8s/v0/vault_client.py:474]: I'm assuming by the way this is written, `issuers_config` exists as some default value even before we write anything?

Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_client.py:474]: Yes, my intention is to wait until Vault creates the config then we modify it.

**Review by DanielArndt — APPROVED** [review]

**Comment by DanielArndt** [comment]
I'm going to merge this because it will help me publish the latest library changes and get them into the machine charm.

**Pull request merged**